subsys/bluetooth/host/conn.c

```diff
@@ -1660,28 +1660,53 @@ int bt_conn_disconnect(struct bt_conn *conn, uint8_t reason)
 
 /* Group Connected BT_CONN only in this */
 #if defined(CONFIG_BT_CONN)
+
 /* We don't want the application to get a PHY update callback upon connection
  * establishment on 2M PHY. Therefore we must prevent issuing LE Set PHY
  * in this scenario.
+ *
+ * It is ifdef'd because the struct fields don't exist in some configs.
  */
-static bool skip_auto_phy_update_on_conn_establishment(struct bt_conn *conn)
+static bool uses_symmetric_2mbit_phy(struct bt_conn *conn)
 {
 #if defined(CONFIG_BT_USER_PHY_UPDATE)
-	if (IS_ENABLED(CONFIG_BT_AUTO_PHY_UPDATE) &&
-	    IS_ENABLED(CONFIG_BT_EXT_ADV) &&
-	    BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
+	if (IS_ENABLED(CONFIG_BT_EXT_ADV)) {
 		if (conn->le.phy.tx_phy == BT_HCI_LE_PHY_2M &&
 		    conn->le.phy.rx_phy == BT_HCI_LE_PHY_2M) {
 			return true;
 		}
 	}
 #else
 	ARG_UNUSED(conn);
-#endif /* defined(CONFIG_BT_USER_PHY_UPDATE) */
+#endif
 
 	return false;
 }
 
+static bool can_initiate_feature_exchange(struct bt_conn *conn)
+{
+	/* Spec says both central and peripheral can send the command. However,
+	 * peripheral-initiated feature exchange is an optional feature.
+	 *
+	 * We provide an optimization if we are in the same image as the
+	 * controller, as we know at compile time whether it supports or not
+	 * peripheral feature exchange.
+	 */
+	bool onboard_controller = IS_ENABLED(CONFIG_BT_CTLR);
+	bool supports_peripheral_feature_exchange = IS_ENABLED(CONFIG_BT_CTLR_PER_INIT_FEAT_XCHG);
```

Comment:
This is perhaps slightly unclean, especially with the new HCI driver model. Even though such a config probably doesn't make much sense, you can in theory enable multiple HCI drivers, but what the host uses ultimately depends on the `zephyr,bt-hci` DT chosen property.

Another thought, unrelated to this PR: we actually have multiple "local controller" implementations in Zephyr which may be part of the same image. The one referred to here just happens to be the only open source one, whereas others get pulled in by binary blobs and respective HCI drivers under `drivers/bluetooth/hci`. I wonder if we could somehow make the controller Kconfig namespace flexible enough that `CONFIG_BT_CTRL` would just mean any local controller, and the platform/HCI driver specific Kconfig for it would set the feature options accordingly.

Anyway, neither of the above points are really change requests to this PR, just something that came to mind when looking at this code :)